[BUG][typescript-fetch] instanceOf type guards produce "TS2590: union type too complex" for models with many sanitized (snake_case) required properties

Bug Report Checklist
- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator?
- [x] Have you tested with the latest master to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?

Description

As of 7.23.0, the typescript-fetch generator emits instanceOf<Model> type guards that fail to compile with error TS2590: Expression produces a union type that is too complex to represent when a model has many required properties whose name differs from baseName (e.g. snake_case JSON keys camelCased into the TS interface).

This is a regression introduced by #23497, which added a dual-key check guarded by hasSanitizedName in modelGeneric.mustache:

{{#hasSanitizedName}}
if ((!('{{name}}' in value) && !('{{baseName}}' in value)) || (value['{{name}}'] === undefined && value['{{baseName}}'] === undefined)) return false;
{{/hasSanitizedName}}

Because the function is a type predicate (value is {{classname}}) and each clause performs two indexed-access reads (value['camelCase'], value['snake_case']) against a parameter typed object, TypeScript computes a narrowed candidate union after every clause. For a model where most/all required fields are sanitized (any snake_case API), the union width grows past the compiler's internal limit and trips TS2590. Skinny models stay under the threshold; wide models (≈15+ sanitized required props) fail.

APIs that serialize snake_case are hit on nearly every model, since name !== baseName for essentially every field.

openapi-generator version

7.23.0 (regression introduced in this version; 7.22.0 and earlier generate compilable guards). Latest release at time of filing.

OpenAPI declaration / minimal spec
```
openapi: 3.0.3
info:
  title: repro
  version: 1.0.0
paths:
  /thing:
    get:
      operationId: getThing
      responses:
        '200':
          description: ok
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Thing'
components:
  schemas:
    Thing:
      type: object
      properties:
        aws_region:        { type: string }
        aws_account_name:  { type: string }
        aws_account_number:{ type: string }
        created_at:        { type: string }
        updated_at:        { type: string }
        # ...repeat enough snake_case required props to cross the union limit (~15+)
      required:
        - aws_region
        - aws_account_name
        - aws_account_number
        - created_at
        - updated_at
```
Generation command
```
openapi-generator-cli generate -i repro.yaml -g typescript-fetch -o ./out
cd out && tsc --noEmit
```
Generated output (abbreviated)
```
export function instanceOfThing(value: object): value is Thing {
    if ((!('awsRegion' in value) && !('aws_region' in value)) || (value['awsRegion'] === undefined && value['aws_region'] === undefined)) return false;
    // ...one per required field...
    return true;
}
```
tsc reports TS2590 on the value['...'] === undefined reads.

Steps to reproduce
1. Generate the spec above with typescript-fetch on 7.23.0.
2. Run tsc --noEmit against the output.
3. Observe TS2590 on the instanceOf guard.

Suggested fix

A few directions (maintainers' call):
- Widen the parameter type or read through (value as any) for the indexed-access checks so TS doesn't narrow the predicate on each clause; or
- Keep the discriminator-value fix from #23497 but drop the dual name/baseName membership narrowing in favor of checking only baseName (the raw-JSON key) — which was the actual bug #23497 targeted; or
- Split the single boolean chain into sequential statements that don't accumulate one large union.

Related issues/PRs

Regression from #23497 (merged in 7.23.0).

--- a/modules/openapi-generator/src/main/resources/typescript-fetch/modelGeneric.mustache
+++ b/modules/openapi-generator/src/main/resources/typescript-fetch/modelGeneric.mustache
@@ -26,7 +26,7 @@ export function instanceOf{{classname}}(value: object): value is {{classname}} {
     {{#vars}}
     {{#required}}
     {{#hasSanitizedName}}
-    if ((!('{{name}}' in value) && !('{{baseName}}' in value)) || (value['{{name}}'] === undefined && value['{{baseName}}'] === undefined)) return false;
+    if ((!('{{name}}' in (value as Record<string, any>)) && !('{{baseName}}' in (value as Record<string, any>))) || ((value as Record<string, any>)['{{name}}'] === undefined && (value as Record<string, any>)['{{baseName}}'] === undefined)) return false;
     {{/hasSanitizedName}}
     {{^hasSanitizedName}}
     if (!('{{name}}' in value) || value['{{name}}'] === undefined) return false;
@@ -37,8 +37,8 @@ export function instanceOf{{classname}}(value: object): value is {{classname}} {
     {{#-first}}
     {{#-last}}
     {{#hasSanitizedName}}
-    {{#isString}}if (value['{{name}}'] !== '{{.}}' && value['{{baseName}}'] !== '{{.}}') return false;{{/isString}}
-    {{^isString}}if (value['{{name}}'] !== {{.}} && value['{{baseName}}'] !== {{.}}) return false;{{/isString}}
+    {{#isString}}if ((value as Record<string, any>)['{{name}}'] !== '{{.}}' && (value as Record<string, any>)['{{baseName}}'] !== '{{.}}') return false;{{/isString}}
+    {{^isString}}if ((value as Record<string, any>)['{{name}}'] !== {{.}} && (value as Record<string, any>)['{{baseName}}'] !== {{.}}) return false;{{/isString}}
     {{/hasSanitizedName}}
     {{^hasSanitizedName}}
     {{#isString}}if (value['{{name}}'] !== '{{.}}') return false;{{/isString}}

--- a/modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/fetch/TypeScriptFetchClientCodegenTest.java
+++ b/modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/fetch/TypeScriptFetchClientCodegenTest.java
@@ -466,25 +466,26 @@ public class TypeScriptFetchClientCodegenTest {
 
         // SnakeOptionOne: discriminator_field (snake_case baseName) vs discriminatorField (camelCase name)
         // instanceOf should check both casings for field presence and discriminator value
+        // (the value is widened to Record<string, any> to avoid TS2590 on wide models)
         Path snakeOptionOne = Paths.get(output + "/models/SnakeOptionOne.ts");
         TestUtils.assertFileExists(snakeOptionOne);
-        TestUtils.assertFileContains(snakeOptionOne, "'discriminatorField' in value");
-        TestUtils.assertFileContains(snakeOptionOne, "'discriminator_field' in value");
-        TestUtils.assertFileContains(snakeOptionOne, "value['discriminatorField'] !== 'snakeOptionOne'");
-        TestUtils.assertFileContains(snakeOptionOne, "value['discriminator_field'] !== 'snakeOptionOne'");
+        TestUtils.assertFileContains(snakeOptionOne, "'discriminatorField' in (value as Record<string, any>)");
+        TestUtils.assertFileContains(snakeOptionOne, "'discriminator_field' in (value as Record<string, any>)");
+        TestUtils.assertFileContains(snakeOptionOne, "(value as Record<string, any>)['discriminatorField'] !== 'snakeOptionOne'");
+        TestUtils.assertFileContains(snakeOptionOne, "(value as Record<string, any>)['discriminator_field'] !== 'snakeOptionOne'");
         // Also verify the non-enum required field checks both casings
-        TestUtils.assertFileContains(snakeOptionOne, "'someProperty' in value");
-        TestUtils.assertFileContains(snakeOptionOne, "'some_property' in value");
+        TestUtils.assertFileContains(snakeOptionOne, "'someProperty' in (value as Record<string, any>)");
+        TestUtils.assertFileContains(snakeOptionOne, "'some_property' in (value as Record<string, any>)");
 
         // DashedOptionOne: discriminator-field (dashed baseName) vs discriminatorField (camelCase name)
         Path dashedOptionOne = Paths.get(output + "/models/DashedOptionOne.ts");
         TestUtils.assertFileExists(dashedOptionOne);
-        TestUtils.assertFileContains(dashedOptionOne, "'discriminatorField' in value");
-        TestUtils.assertFileContains(dashedOptionOne, "'discriminator-field' in value");
-        TestUtils.assertFileContains(dashedOptionOne, "value['discriminatorField'] !== 'dashedOptionOne'");
-        TestUtils.assertFileContains(dashedOptionOne, "value['discriminator-field'] !== 'dashedOptionOne'");
-        TestUtils.assertFileContains(dashedOptionOne, "'someProperty' in value");
-        TestUtils.assertFileContains(dashedOptionOne, "'some-property' in value");
+        TestUtils.assertFileContains(dashedOptionOne, "'discriminatorField' in (value as Record<string, any>)");
+        TestUtils.assertFileContains(dashedOptionOne, "'discriminator-field' in (value as Record<string, any>)");
+        TestUtils.assertFileContains(dashedOptionOne, "(value as Record<string, any>)['discriminatorField'] !== 'dashedOptionOne'");
+        TestUtils.assertFileContains(dashedOptionOne, "(value as Record<string, any>)['discriminator-field'] !== 'dashedOptionOne'");
+        TestUtils.assertFileContains(dashedOptionOne, "'someProperty' in (value as Record<string, any>)");
+        TestUtils.assertFileContains(dashedOptionOne, "'some-property' in (value as Record<string, any>)");
 
         // Numeric singleton enum: value check must NOT quote the literal
         Path numericModel = Paths.get(output + "/models/NumericSingletonEnumModel.ts");

--- a/samples/client/petstore/typescript-fetch/builds/default-v3.0/models/EnumTest.ts
+++ b/samples/client/petstore/typescript-fetch/builds/default-v3.0/models/EnumTest.ts
@@ -142,7 +142,7 @@ export type EnumTestEnumNumberEnum = typeof EnumTestEnumNumberEnum[keyof typeof 
  * Check if a given object implements the EnumTest interface.
  */
 export function instanceOfEnumTest(value: object): value is EnumTest {
-    if ((!('enumStringRequired' in value) && !('enum_string_required' in value)) || (value['enumStringRequired'] === undefined && value['enum_string_required'] === undefined)) return false;
+    if ((!('enumStringRequired' in (value as Record<string, any>)) && !('enum_string_required' in (value as Record<string, any>))) || ((value as Record<string, any>)['enumStringRequired'] === undefined && (value as Record<string, any>)['enum_string_required'] === undefined)) return false;
     return true;
 }
 

--- a/samples/client/petstore/typescript-fetch/builds/default-v3.0/models/FormatTest.ts
+++ b/samples/client/petstore/typescript-fetch/builds/default-v3.0/models/FormatTest.ts
@@ -122,7 +122,7 @@ export interface FormatTest {
  */
 export function instanceOfFormatTest(value: object): value is FormatTest {
     if (!('number' in value) || value['number'] === undefined) return false;
-    if ((!('_byte' in value) && !('byte' in value)) || (value['_byte'] === undefined && value['byte'] === undefined)) return false;
+    if ((!('_byte' in (value as Record<string, any>)) && !('byte' in (value as Record<string, any>))) || ((value as Record<string, any>)['_byte'] === undefined && (value as Record<string, any>)['byte'] === undefined)) return false;
     if (!('date' in value) || value['date'] === undefined) return false;
     if (!('password' in value) || value['password'] === undefined) return false;
     return true;

--- a/samples/client/petstore/typescript-fetch/builds/oneOf/models/DashedOptionOne.ts
+++ b/samples/client/petstore/typescript-fetch/builds/oneOf/models/DashedOptionOne.ts
@@ -47,10 +47,10 @@ export type DashedOptionOneDiscriminatorFieldEnum = typeof DashedOptionOneDiscri
  * Check if a given object implements the DashedOptionOne interface.
  */
 export function instanceOfDashedOptionOne(value: object): value is DashedOptionOne {
-    if ((!('discriminatorField' in value) && !('discriminator-field' in value)) || (value['discriminatorField'] === undefined && value['discriminator-field'] === undefined)) return false;
-    if (value['discriminatorField'] !== 'dashedOptionOne' && value['discriminator-field'] !== 'dashedOptionOne') return false;
+    if ((!('discriminatorField' in (value as Record<string, any>)) && !('discriminator-field' in (value as Record<string, any>))) || ((value as Record<string, any>)['discriminatorField'] === undefined && (value as Record<string, any>)['discriminator-field'] === undefined)) return false;
+    if ((value as Record<string, any>)['discriminatorField'] !== 'dashedOptionOne' && (value as Record<string, any>)['discriminator-field'] !== 'dashedOptionOne') return false;
     
-    if ((!('someProperty' in value) && !('some-property' in value)) || (value['someProperty'] === undefined && value['some-property'] === undefined)) return false;
+    if ((!('someProperty' in (value as Record<string, any>)) && !('some-property' in (value as Record<string, any>))) || ((value as Record<string, any>)['someProperty'] === undefined && (value as Record<string, any>)['some-property'] === undefined)) return false;
     return true;
 }
 

--- a/samples/client/petstore/typescript-fetch/builds/oneOf/models/DashedOptionTwo.ts
+++ b/samples/client/petstore/typescript-fetch/builds/oneOf/models/DashedOptionTwo.ts
@@ -47,10 +47,10 @@ export type DashedOptionTwoDiscriminatorFieldEnum = typeof DashedOptionTwoDiscri
  * Check if a given object implements the DashedOptionTwo interface.
  */
 export function instanceOfDashedOptionTwo(value: object): value is DashedOptionTwo {
-    if ((!('discriminatorField' in value) && !('discriminator-field' in value)) || (value['discriminatorField'] === undefined && value['discriminator-field'] === undefined)) return false;
-    if (value['discriminatorField'] !== 'dashedOptionTwo' && value['discriminator-field'] !== 'dashedOptionTwo') return false;
+    if ((!('discriminatorField' in (value as Record<string, any>)) && !('discriminator-field' in (value as Record<string, any>))) || ((value as Record<string, any>)['discriminatorField'] === undefined && (value as Record<string, any>)['discriminator-field'] === undefined)) return false;
+    if ((value as Record<string, any>)['discriminatorField'] !== 'dashedOptionTwo' && (value as Record<string, any>)['discriminator-field'] !== 'dashedOptionTwo') return false;
     
-    if ((!('someProperty' in value) && !('some-property' in value)) || (value['someProperty'] === undefined && value['some-property'] === undefined)) return false;
+    if ((!('someProperty' in (value as Record<string, any>)) && !('some-property' in (value as Record<string, any>))) || ((value as Record<string, any>)['someProperty'] === undefined && (value as Record<string, any>)['some-property'] === undefined)) return false;
     return true;
 }
 

--- a/samples/client/petstore/typescript-fetch/builds/oneOf/models/SnakeOptionOne.ts
+++ b/samples/client/petstore/typescript-fetch/builds/oneOf/models/SnakeOptionOne.ts
@@ -47,10 +47,10 @@ export type SnakeOptionOneDiscriminatorFieldEnum = typeof SnakeOptionOneDiscrimi
  * Check if a given object implements the SnakeOptionOne interface.
  */
 export function instanceOfSnakeOptionOne(value: object): value is SnakeOptionOne {
-    if ((!('discriminatorField' in value) && !('discriminator_field' in value)) || (value['discriminatorField'] === undefined && value['discriminator_field'] === undefined)) return false;
-    if (value['discriminatorField'] !== 'snakeOptionOne' && value['discriminator_field'] !== 'snakeOptionOne') return false;
+    if ((!('discriminatorField' in (value as Record<string, any>)) && !('discriminator_field' in (value as Record<string, any>))) || ((value as Record<string, any>)['discriminatorField'] === undefined && (value as Record<string, any>)['discriminator_field'] === undefined)) return false;
+    if ((value as Record<string, any>)['discriminatorField'] !== 'snakeOptionOne' && (value as Record<string, any>)['discriminator_field'] !== 'snakeOptionOne') return false;
     
-    if ((!('someProperty' in value) && !('some_property' in value)) || (value['someProperty'] === undefined && value['some_property'] === undefined)) return false;
+    if ((!('someProperty' in (value as Record<string, any>)) && !('some_property' in (value as Record<string, any>))) || ((value as Record<string, any>)['someProperty'] === undefined && (value as Record<string, any>)['some_property'] === undefined)) return false;
     return true;
 }
 

--- a/samples/client/petstore/typescript-fetch/builds/oneOf/models/SnakeOptionTwo.ts
+++ b/samples/client/petstore/typescript-fetch/builds/oneOf/models/SnakeOptionTwo.ts
@@ -47,10 +47,10 @@ export type SnakeOptionTwoDiscriminatorFieldEnum = typeof SnakeOptionTwoDiscrimi
  * Check if a given object implements the SnakeOptionTwo interface.
  */
 export function instanceOfSnakeOptionTwo(value: object): value is SnakeOptionTwo {
-    if ((!('discriminatorField' in value) && !('discriminator_field' in value)) || (value['discriminatorField'] === undefined && value['discriminator_field'] === undefined)) return false;
-    if (value['discriminatorField'] !== 'snakeOptionTwo' && value['discriminator_field'] !== 'snakeOptionTwo') return false;
+    if ((!('discriminatorField' in (value as Record<string, any>)) && !('discriminator_field' in (value as Record<string, any>))) || ((value as Record<string, any>)['discriminatorField'] === undefined && (value as Record<string, any>)['discriminator_field'] === undefined)) return false;
+    if ((value as Record<string, any>)['discriminatorField'] !== 'snakeOptionTwo' && (value as Record<string, any>)['discriminator_field'] !== 'snakeOptionTwo') return false;
     
-    if ((!('someProperty' in value) && !('some_property' in value)) || (value['someProperty'] === undefined && value['some_property'] === undefined)) return false;
+    if ((!('someProperty' in (value as Record<string, any>)) && !('some_property' in (value as Record<string, any>))) || ((value as Record<string, any>)['someProperty'] === undefined && (value as Record<string, any>)['some_property'] === undefined)) return false;
     return true;
 }
 

--- a/samples/client/petstore/typescript-fetch/builds/snakecase-discriminator/models/Animal.ts
+++ b/samples/client/petstore/typescript-fetch/builds/snakecase-discriminator/models/Animal.ts
@@ -39,7 +39,7 @@ export interface Animal {
  * Check if a given object implements the Animal interface.
  */
 export function instanceOfAnimal(value: object): value is Animal {
-    if ((!('className' in value) && !('class_name' in value)) || (value['className'] === undefined && value['class_name'] === undefined)) return false;
+    if ((!('className' in (value as Record<string, any>)) && !('class_name' in (value as Record<string, any>))) || ((value as Record<string, any>)['className'] === undefined && (value as Record<string, any>)['class_name'] === undefined)) return false;
     return true;
 }
 

--- a/samples/client/petstore/typescript-fetch/builds/snakecase-discriminator/models/EnumTest.ts
+++ b/samples/client/petstore/typescript-fetch/builds/snakecase-discriminator/models/EnumTest.ts
@@ -142,7 +142,7 @@ export type EnumTestEnumNumberEnum = typeof EnumTestEnumNumberEnum[keyof typeof 
  * Check if a given object implements the EnumTest interface.
  */
 export function instanceOfEnumTest(value: object): value is EnumTest {
-    if ((!('enumStringRequired' in value) && !('enum_string_required' in value)) || (value['enumStringRequired'] === undefined && value['enum_string_required'] === undefined)) return false;
+    if ((!('enumStringRequired' in (value as Record<string, any>)) && !('enum_string_required' in (value as Record<string, any>))) || ((value as Record<string, any>)['enumStringRequired'] === undefined && (value as Record<string, any>)['enum_string_required'] === undefined)) return false;
     return true;
 }
 

--- a/samples/client/petstore/typescript-fetch/builds/snakecase-discriminator/models/FormatTest.ts
+++ b/samples/client/petstore/typescript-fetch/builds/snakecase-discriminator/models/FormatTest.ts
@@ -122,7 +122,7 @@ export interface FormatTest {
  */
 export function instanceOfFormatTest(value: object): value is FormatTest {
     if (!('number' in value) || value['number'] === undefined) return false;
-    if ((!('_byte' in value) && !('byte' in value)) || (value['_byte'] === undefined && value['byte'] === undefined)) return false;
+    if ((!('_byte' in (value as Record<string, any>)) && !('byte' in (value as Record<string, any>))) || ((value as Record<string, any>)['_byte'] === undefined && (value as Record<string, any>)['byte'] === undefined)) return false;
     if (!('date' in value) || value['date'] === undefined) return false;
     if (!('password' in value) || value['password'] === undefined) return false;
     return true;

